agents/go-agents/core/event/event.go

```diff
@@ -81,19 +82,20 @@ func (bus *Bus) Pub(e E) {
 func (bus *Bus) Sub(consumer Consumer, eType string) {
 	bus.Lock()
 	defer bus.Unlock()
-	bus.consumers[eType] = append(bus.consumers[eType], consumer)
+	bus.consumers[eType] = append(bus.copyConsumers(eType), consumer)
```

Comment:
Since the whole function is covered by a lock do we really need to copy consumers before appending the new one?

Reply:
Technically it will work without copying, as a new element will be added to the underlining array while concurrently reading(publishing) go routines will be working with slices referencing the same array but different elements.

The reason i changed this code was to achieve the same copy-on-write pattern concept throughout all bus state modification functions. For example `Bus.rm` copies values because otherwise it would change the original values of underlining array, including those being concurrently referenced by reading routines, which may produce unexpected results. _Fail-safe_ traversing, simplicity are the main reasons. 

If we take a look at `EventService` java implementation it carries `ConcurrentMap`of `CopyOnWriteArraySets`, so each write produces a new instance of underlining array, which is the same approach in case of modifications.

Reply:
I thought that it always creates a new array, but now I see that there are cases when the slice is just resliced. 
Thank you for the explanation.